### ISSUE TITLE
Possible SIMD intrinsics bug on System.Numerics Vector types

### ISSUE THREAD

**Issue by TIHan** [dotnet/runtime#35432]
The following F# code, when run in .NET Core 3.0 or 3.1, will hit the `failwith`:
```fsharp
open System.Numerics

[<Struct>]
type Vertex(position: Vector3, uv: Vector2) =

    member _.Position = position

    member _.Uv = uv

let test len =

    let positions = Array.init len (fun i -> Vector3(single i, single i, single i))
    let uv = Array.init len (fun i -> Vector2(single i, single i))

    let vertices = Array.init len (fun i -> Vertex(positions.[i], uv.[i]))
    if vertices.[2].Position <> Vector3(2.f, 2.f, 2.f) then
        failwith "Failed"

[<EntryPoint>]
let main argv =
    test 10
    0
```

This is compiled with VS 16.5.1's F# compiler as a Release build. Turning off tail calls doesn't make a difference either, so it's not a tail call issue

`vertices.[2]` should be a `Vector3` with each component's value being `2.f`. Instead, its zero'd out. In fact, all values in `vertices` are zero'd out which they should not be.

This only happens when I run the compiled assembly of this F# code in .NET Core 3.0 or 3.1. In .NET Core 2.1 and below, this passes.

Also, defining my own `Vector3` type will cause this to pass in .NET Core 3.0/3.1:
```fsharp
open System.Numerics

type Vector3(x: single, y: single, z: single) = struct end

[<Struct>]
type Vertex(position: Vector3, uv: Vector2) =

    member _.Position = position

    member _.Uv = uv

let test len =

    let positions = Array.init len (fun i -> Vector3(single i, single i, single i))
    let uv = Array.init len (fun i -> Vector2(single i, single i))

    let vertices = Array.init len (fun i -> Vertex(positions.[i], uv.[i]))
    if vertices.[2].Position <> Vector3(2.f, 2.f, 2.f) then
        failwith "Failed"

[<EntryPoint>]
let main argv =
    test 10
    0

```

**Comment by tannergooding** [comment]
Looks to be related to the SIMD Intrinsics, not necessarily HWIntrinsics. I was not able to reproduce using "equivalent" C# code and the only notable difference in IL is that F# generates a few tail calls.

CC. @CarolEidt, @echesakovMSFT. Also possibly @AndyAyersMS

**Comment by tannergooding** [comment]
IL from sharplab: https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AbEAzAzmgFxAEsMAfCABxgDsACAZQE9cCYBbAOgDkBXdmFGJhcAWABQEgk2p0AajDAFoAZgAUCEHVzEaAcwww0dJlp37DxgF5ndBmAEo6AXm0EovJXVoATCRIBtAB4Gd08CAD4AXSkZGHlBNgQ1SggdAmIIGi0FJVVjXgA3HMVlKAAmJ2d/cTo6ugF2YEE6AH1OAAU04gyslzpU9MyaGvqGjmaoNs4AVUL+oprDAjo2VjpDemrJWvrlge7emlx+gEEoKABDJk5dHo3aOjVsXnpiOjgIhLyodXN7OjEYz/QyA4F2UHEBwOCRjfZFM4Xa63Gj3TZPF5vD5fXJlcpqEHxIHaCFE6GjPYwFaFRLCGAnVznK43O4rdHPV6A7EJKBJFKHYa4TgBYhRAqFYWi8m7OrEbB0Gm8ulCgLlKKdAV9II40qqNTlTjYYwGo10E1OAgAC1osLG9Wwl1IAHcepa6AAiABijsMPndNWCAFEaO4mF1dJEYuJ9uxHfRLlA9PNtmM1isAIwABltdUzQA=

**Comment by TIHan** [comment]
Sharplab wont give you the ability to turn off tail calls, but you can turn them off in VS:
![Untitled](https://user-images.githubusercontent.com/1278959/80266632-c8f20a80-8651-11ea-80e5-b3aeb67b2fdb.png)

**Comment by AndyAyersMS** [comment]
@TIHan which 3.1 version are you using?

I wonder if this is something that we've already fixed in servicing -- I have a pre-release version of 3.1.4 installed on my box, and the test passes there.

**Comment by TIHan** [comment]
I was using .NET Core 3.1.2. But, I just installed .NET Core 3.1.3, and the issue doesn't happen. So, looks like it's fixed ;)

**Comment by AndyAyersMS** [comment]
Seems like this could be an instance of #1241 which was fixed for 3.1 via dotnet/coreclr#27983.

If so this would have landed in 3.1.3, so matches well with your observations.

**Comment by TIHan** [comment]
Yup, I just stumbled onto that just now and I believe that was it.
